packages/react-dom/src/client/ReactDOMHostConfig.js

```diff
@@ -1013,14 +1012,84 @@ export function getInstanceFromNode(node: HTMLElement): null | Object {
   return getClosestInstanceFromNode(node) || null;
 }
 
+function makeNestedClientId(
+  parent: OpaqueIDType,
+  getValueOfKey: () => string,
+  depth: number,
+): OpaqueIDType {
+  return {
+    makeNewFromKey(key: string) {
+      return makeNestedClientId(this, () => key, depth + 1);
+    },
+    toString() {
+      //$FlowFixMe toString is only ever void if attemptToReadValue throws, which means we don't get here anyway
+      return parent.toString() + getValueOfKey() + ':' + depth.toString(32);
+    },
+    valueOf() {
+      //$FlowFixMe toString is only ever void if attemptToReadValue throws, which means we don't get here anyway
+      return parent.valueOf() + getValueOfKey() + ':' + depth.toString(32);
+    },
+  };
+}
+
 let clientId: number = 0;
 export function makeClientId(): OpaqueIDType {
-  return 'r:' + (clientId++).toString(36);
+  const id = 'r:' + (clientId++).toString(36);
+  return {
+    makeNewFromKey(key) {
```

Comment:
Should return types be added on these?

Reply:
OpaqueIdType already has the return types of these functions